src/controllers/krate/versions.rs

```diff
@@ -21,27 +27,217 @@ pub async fn versions(state: AppState, Path(crate_name): Path<String>) -> AppRes
+        // Sorting by semver but opted for id as the seek key because num can be quite lengthy,
+        // while id values are significantly smaller.
+        let mut sorted_versions = IndexMap::new();
+        for result in krate
+            .all_versions()
+            .select((versions::id, versions::num))
+            .load_iter::<(i32, String), DefaultLoadingMode>(conn)?
+        {
+            let (id, num) = result?;
+            sorted_versions.insert(id, (num, None));
+        }
+        sorted_versions.sort_by_cached_key(|_, (num, _)| Reverse(semver::Version::parse(num).ok()));
+
+        let mut idx = Some(0);
+        if let Page::Seek(ref seek) = options.page {
+            idx = seek
+                .decode::<i32>()
+                .map(|id| sorted_versions.get_index_of(&id))?
+                .filter(|i| i + 1 < sorted_versions.len())
+                .map(|i| i + 1);
+        }
+        if let Some(start) = idx {
+            let end = (start + options.per_page as usize).min(sorted_versions.len());
+            let ids = sorted_versions[start..end].keys().collect::<Vec<_>>();
+            for result in krate
+                .all_versions()
+                .left_outer_join(users::table)
+                .select((versions::all_columns, users::all_columns.nullable()))
+                .filter(versions::id.eq_any(ids))
+                .load_iter::<(Version, Option<User>), DefaultLoadingMode>(conn)?
+            {
+                let row = result?;
+                sorted_versions.insert(row.0.id, (row.0.num.to_owned(), Some(row)));
+            }
+            (
+                sorted_versions
+                    .values()
+                    .flat_map(|(_, v)| v)
+                    .cloned()
+                    .collect(),
+                sorted_versions.len(),
+            )
+        } else {
+            (vec![], 0)
+        }
+    } else {
+        let mut data: Vec<(Version, Option<User>)> = krate
+            .all_versions()
+            .left_outer_join(users::table)
+            .select((versions::all_columns, users::all_columns.nullable()))
+            .load(conn)?;
+        data.sort_by_cached_key(|(version, _)| Reverse(semver::Version::parse(&version.num).ok()));
+        let total = data.len();
+        (data, total)
+    };
+
+    let mut next_page = None;
+    if let Some(options) = options {
+        next_page =
+            next_seek_params(&data, options, |last| last.0.id)?.map(|p| req.query_with_params(p))
+    };
+
+    Ok(PaginatedVersionsAndPublishers {
+        data,
+        meta: ResponseMeta {
+            total: total as i64,
+            next_page,
+        },
+    })
+}
+
+mod seek {
+    use chrono::naive::serde::ts_microseconds;
+    use serde::{Deserialize, Serialize};
+
+    #[derive(Deserialize, Serialize)]
+    pub(super) struct Date(
+        #[serde(with = "ts_microseconds")] pub(super) chrono::NaiveDateTime,
+        pub(super) i32,
+    );
```

Comment:
I think I'd feel more comfortable with named fields for this, unless there is a strong reason why we can't use that.

Reply:
The reason for this is to align the implementation of seek pagination with other routes. Currently, we encode seek as a `Vec`, which doesn't include field names. While we could still customize the `Serializer/Deserializer` for it, I think it would be better to implement this alignment in a separate PR.

Also, #7983 has already been merged, so I would add another fixup commit to use the `seek!` macro to further align the implementation.

Reply:
hmm okay, sounds like there are valid reasons. might be good to add write them down in a code comment though :)